### PR TITLE
Fix: XSS vulnerability on Woo Checkout user input

### PR DESCRIPTION
**Pull request by alexmigf**

### CODE REVIEW

File: includes/wcpdf-functions.php [hunk 1]
```diff
@@ -489,6 +489,97 @@ function wcpdf_convert_encoding( $string, $tool = 'mb_convert_encoding' ) {
...
+ *
+ * @return string
+ */
+function wpo_wcpdf_sanitize_html_content( string $html, string $context = '', array $allowed_tags = array() ): string {
+	if ( empty( $html ) ) {
+		return $html;
+	}
+
+	// default allowed tags
+	$allowed_tags = array_merge( apply_filters( 'wpo_wcpdf_sanitize_html_default_allowed_tags', array(
```
Comment (MohamadNateqi) [inline includes/wcpdf-functions.php:507]: Do you think we should be concerned about the filter here?

**Review by MohamadNateqi — CHANGES_REQUESTED** [review]

File: includes/wcpdf-functions.php [hunk 2]
```diff
@@ -489,6 +489,97 @@ function wcpdf_convert_encoding( $string, $tool = 'mb_convert_encoding' ) {
...
+	foreach ( $xpath->query( '//*' ) as $node ) {
+		// check if the node is allowed.
+		if ( array_key_exists( $node->nodeName, $allowed_tags ) ) {
+			// if the node is allowed, check each attribute.
+			$attributes_to_remove = [];
+			
+			foreach ( $node->attributes as $attr ) {
+				if ( ! in_array( $attr->nodeName, $allowed_tags[ $node->nodeName ] ) ) {
+					// mark the attribute for removal if it's not in the list of allowed attributes.
+					$attributes_to_remove[] = $attr->nodeName;
```
[the remark this answers is not in the document]
Comment (alexmigf) [reply includes/wcpdf-functions.php]: My bad!

File: includes/documents/abstract-wcpdf-order-document-methods.php [hunk 3]
```diff
@@ -248,15 +252,30 @@ public function get_custom_field( $field_name ) {
...
+				'img'  => array( 'src', 'title', 'alt' ),
+				'p'    => array(),
+				'span' => array(),
+				'ul'   => array(),
+				'ol'   => array(),
+				'li'   => array(),
+			);
+			
 			if ( is_array( $custom_field ) ) {
+				$custom_field = array_map( function( $field ) {
```
[the remark this answers is not in the document]
Comment (alexmigf) [reply includes/documents/abstract-wcpdf-order-document-methods.php]: Forgot that, thank you for catching it.

**Comment by alexmigf** [comment]
@MohamadNateqi thank you for your precious testing!

File: includes/documents/abstract-wcpdf-order-document-methods.php [hunk 4]
```diff
@@ -82,21 +82,23 @@ public function get_billing_address() {
...
+			unset( $refund );
 		} else {
+			// regular shop_order
+			$address = $this->order->get_formatted_billing_address();
+			
 			// no address
-			$address = apply_filters( 'wpo_wcpdf_billing_address', __('N/A', 'woocommerce-pdf-invoices-packing-slips' ), $this );
+			if ( empty( $address ) ) {
+				$address = __( 'N/A', 'woocommerce-pdf-invoices-packing-slips' );
+			}
```
Comment (Terminator-Barbapapa) [inline includes/documents/abstract-wcpdf-order-document-methods.php:98]: Shouldn't we put this outside the if/else statement, so it also works for refunds with a parent order that has no address?

File: includes/documents/abstract-wcpdf-order-document-methods.php [hunk 5]
```diff
@@ -176,27 +179,28 @@ public function get_shipping_address() {
...
-			} else{
-				$address = apply_filters( 'wpo_wcpdf_shipping_address', __('N/A', 'woocommerce-pdf-invoices-packing-slips' ), $this );
-
+			if ( empty( $address ) ) {
+				// use fallback for packing slip
+				if ( apply_filters( 'wpo_wcpdf_shipping_address_fallback', ( 'packing-slip' === $this->get_type() ), $this ) ) {
+					$address = $this->get_billing_address();
+				} else{
+					$address = __( 'N/A', 'woocommerce-pdf-invoices-packing-slips' );
+				}
```
Comment (Terminator-Barbapapa) [inline includes/documents/abstract-wcpdf-order-document-methods.php:199]: Same here.

**Review by Terminator-Barbapapa — APPROVED** [review]

**Pull request merged**